src/coronawhy_vt/utils/local_processing.py

```diff
@@ -0,0 +1,338 @@
+    :param max_num_tokens: Maximum number of tokens in the vocab entry
+
+    :return (names, ngram_map): (Original tokens with spaces instead of underscores
+        {names: names_with_space_repaced_by_underscores})
+    """
+    ngram_map = {}
+    names = []
+    all_names = []
+
+    with open(filename) as f:
+        for line in f:
+            all_names.append(line.rstrip())
+
+    for name in all_names:
+        if len(name) >= min_word_len:
+            if " " in name:
+                if min_num_tokens <= len(name.split(" ")) <= max_num_tokens:
+                    newname = name.replace(" ", "_").lower()
+                    ngram_map[name.lower()] = newname
+                    names.append(newname)
+            else:
+                names.append(name.lower())
+    names = np.unique(names)
+    return names, ngram_map
+
+
+# TODO Pull the next 6 functions into a class?
+# These functions determine what blocks are pulled from the paper for matching
+def title_blocks(paper: dict) -> list:
+    """Retrieve title block of a paper in json form.
+
+    :param paper (json): The paper to be parsed
+
+    :return: Lines of text in the paper title
+    """
+    return [{'text': paper['metadata']['title']}]
+
+
+def abstract_blocks(paper: dict) -> list:
+    """Retrieve abstract block of a paper in json form.
+
+    :param paper: The paper to be parsed
+
+    :return: Lines of text in the paper abstract
+    """
+    return paper['abstract']
+
+
+def body_blocks(paper: dict) -> list:
+    """Retrieve body block of a paper in json form.
+
+    :param paper: The paper to be parsed
+
+    :return: Lines of text in the paper body
+    """
+    return paper['body_text']
+
+
+# TODO the reference nlp is an external dependence, either add into function or make it a class
+nlp = spacy.load('en_core_web_lg')
+
+
+def pull_mentions(
+    paths: list,
+    block_selector: Callable[[dict], list],
+    sec_name: str,
+    words: list,
+    replace_dict: dict = None,
+    lemmatized: bool = True,
+) -> dict:
+    """Aggregate the positional features of the lemmatized text corpus.
+
+    :param paths: pathlib.Path objects corresponding to directories containing
+        corpus to aggregate features from.
+    :param block_selector: The function to retrieve the relevant block from text
+    :param sec_name: Corresponds with block_selector to select the relevant
+        block of text. Options are 'title', 'abstract', 'body'
+        TODO seems like this should be pulled into the function as it depends
+        entirely on the block_selector argument
+    :param words: List of lemmatized words
+    :param replace_dict: Takes a dict of {token: lemmatized underscore token}
+        and replaces occurences of token in text with its value
+    :param lemmatized: True if lemmatize text corpus first
+
+    :return: [('identifier feature': occurrence)]
+    """
+    Feature = namedtuple('Feature', [
```

Comment:
if you make the `Feature` class module-level instead of being inside the function, it might make it easier to do type annotations and for the IDE to make type inference

Reply:
Would you recommend creating a new file for it or just putting it at the top of local_processing?

Reply:
just at the top of local_processing is fine